lib/dry/schema/messages/template.rb

```diff
@@ -0,0 +1,66 @@
+require 'dry/equalizer'
+
+require 'dry/schema/constants'
+
+module Dry
+  module Schema
+    module Messages
+      # Template wraps a string with interpolation tokens and defines evaluator function
+      # dynamically
+      #
+      # @api private
+      class Template
+        include Dry::Equalizer(:text)
+
+        TOKEN_REGEXP = /%{(\w+)}/
+
+        # !@attribute [r] text
+        #   @return [String]
+        attr_reader :text
+
+        # !@attribute [r] tokens
+        #   @return [Hash]
+        attr_reader :tokens
+
+        # !@attribute [r] evaluator
+        #   @return [Proc]
+        attr_reader :evaluator
+
+        # @api private
+        def self.[](input)
+          new(*parse(input))
+        end
+
+        # @api private
+        def self.parse(input)
+          tokens = input.scan(TOKEN_REGEXP).flatten(1).map(&:to_sym)
+          text = input.gsub('%', '#')
+
+          evaluator = <<-RUBY.strip
+            -> (#{tokens.map { |token| "#{token}:" }.join(", ")}) { "#{text}" }
+          RUBY
+
+          [text, tokens, eval(evaluator)]
```

Comment:
you should pass `__FILE__` and `__LINE__ - 3` here so that it'll show up in backtraces. It's a lesser issue here but it's better to do it by default